source/adapters/level_zero/event.cpp

```diff
@@ -463,6 +465,70 @@ UR_APIEXPORT ur_result_t UR_APICALL urEventGetProfilingInfo(
 
   ze_kernel_timestamp_result_t tsResult;
 
+  // A Command-buffer consists of three command-lists for which only a single
+  // event is returned to users. The actual profiling information related to the
+  // command-buffer should therefore be extrated from graph events themsleves.
+  // The timestamps of these events are saved in a memory region attached to
+  // event usning CommandData field. The timings must therefore be recovered
+  // from this memory.
+  if (Event->CommandType == UR_COMMAND_COMMAND_BUFFER_ENQUEUE_EXP) {
+    if (Event->CommandData) {
+      command_buffer_profiling_t *ProfilingsPtr;
+      switch (PropName) {
+      case UR_PROFILING_INFO_COMMAND_START: {
+        ProfilingsPtr =
+            static_cast<command_buffer_profiling_t *>(Event->CommandData);
+        // Sync-point order does not necessarily match to the order of
+        // execution. We therefore look for the first command executed.
+        uint64_t MinStart = ProfilingsPtr->Timestamps[0].global.kernelStart;
+        for (uint64_t i = 1; i < ProfilingsPtr->NumEvents; i++) {
+          uint64_t Timestamp = ProfilingsPtr->Timestamps[i].global.kernelStart;
+          if (Timestamp < MinStart) {
+            MinStart = Timestamp;
+          }
+        }
+        uint64_t ContextStartTime =
+            (MinStart & TimestampMaxValue) * ZeTimerResolution;
+        return ReturnValue(ContextStartTime);
+      }
+      case UR_PROFILING_INFO_COMMAND_END: {
+        ProfilingsPtr =
+            static_cast<command_buffer_profiling_t *>(Event->CommandData);
+        // Sync-point order does not necessarily match to the order of
+        // execution. We therefore look for the last command executed.
+        uint64_t MaxEnd = ProfilingsPtr->Timestamps[0].global.kernelEnd;
+        uint64_t LastStart = ProfilingsPtr->Timestamps[0].global.kernelStart;
+        for (uint64_t i = 1; i < ProfilingsPtr->NumEvents; i++) {
+          uint64_t Timestamp = ProfilingsPtr->Timestamps[i].global.kernelEnd;
```

Comment:
I dont follow this logic for MaxEnd and the same for MinStart in line 487, could you elaborate?

Reply:
Since we are relaxing the command execution order in a command list (https://github.com/oneapi-src/unified-runtime/pull/1035/files#diff-ab021f9451ccb6dd66e3df853d0e123b2f898233a1f609b1e698bef4f6011c71R397), there is no guarantee that the order of commands in list (order of `SyncPoints`) is the actual execution order.
We thus loop over all the event timestamps of CommandBuffer commands and search for the minimum start time, which corresponds to the actual start time of the CommandBuffer execution. And the same for the execution end time, we go through the events and look for the last task to complete (max `kernelEnd` timestamp).

Reply:
thanks @mfrancepillois . I understand now. main problem there is if the number of commands is very huge, so the timestamp traversal would be substantial. 

Now, the relaxation of ordering you are referring to is because you are adding `ZE_COMMAND_LIST_FLAG_RELAXED_ORDERING`, right? in that case, I don't think the application (in this case UR) should be doing the traversal of timestamps to ordering them. It should be the driver. However the L0 spec is not clear on that.

I have opened an issue in the L0 spec to clarify that https://github.com/oneapi-src/level-zero-spec/issues/249

In the meantime, let's continue with this patch, and then once the L0 spec is clarified, we know whether this code can stay or needs to be removed.